File: java/house/src/main/java/House.java
import java.util.ArrayList;
import java.util.List;

public class House{

    private List<Item> items;

    public House() {
        items = new ArrayList<>();
        items.add(new Item("", ""));
        items.add(new Item("", ""));
        items.add(new Item("the malt", "that lay in the house that Jack built."));
        items.add(new Item("the rat", "ate"));
        items.add(new Item("the cat", "killed"));
        items.add(new Item("the dog", "worried"));
        items.add(new Item("the cow with the crumpled horn", "tossed"));
        items.add(new Item("the maiden all forlorn", "milked"));
        items.add(new Item("the man all tattered and torn", "kissed"));
        items.add(new Item("the priest all shaven and shorn", "married"));
        items.add(new Item("the rooster that crowed in the morn", "woke"));
        items.add(new Item("the farmer sowing his corn", "kept"));
        items.add(new Item("the horse and the hound and the horn", "belonged to"));
    }

    public String verse(int number) {

        if (number == 1) {
            return "This is the house that Jack built.";
        }

        return "This is " + recVerse(number);
        
    }

    public String verses(int startVerse, int endVerse) {
        StringBuilder sb = new StringBuilder();

        for (int i = startVerse; i <= endVerse; i++) {
            sb.append(verse(i)).append("\n");
        }

        return sb.toString().trim();
    }

    public String sing() {
        StringBuilder sb = new StringBuilder();
        sb.append(verse(1)).append("\n");

        sb.append(verses(2, this.items.size() - 1));

        return sb.toString();
    }

    public String recVerse(int number) {
        if(number == 2){
            return items.get(number).getName() + " " + items.get(number).getAction();
        }

        return items.get(number).getName() + " that " + items.get(number).getAction() + " " + recVerse(number - 1);
    }
}

File: javascript/house/house.js
export class House {

    static items = [];

    static {
        this.items.push({name: '', action: ''});
        this.items.push({name: '', action: ''});
        this.items.push({name: 'the malt', 
            action: 'that lay in the house that Jack built.'});
        this.items.push({name: 'the rat', action: 'ate'});
        this.items.push({name: 'the cat', action: 'killed'});
        this.items.push({name: 'the dog', action: 'worried'});
        this.items.push({name: 'the cow with the crumpled horn', action: 'tossed'});
        this.items.push({name: 'the maiden all forlorn', action: 'milked'});
        this.items.push({name: 'the man all tattered and torn', action: 'kissed'});
        this.items.push({name: 'the priest all shaven and shorn', 
            action: 'married'});
        this.items.push({name: 'the rooster that crowed in the morn', 
            action: 'woke'});
        this.items.push({name: 'the farmer sowing his corn', action: 'kept'});
        this.items.push({name: 'the horse and the hound and the horn', 
            action: 'belonged to'});
    }

    static verse(verseNumber) {

        if (verseNumber === 1) {
            return ['This is the house that Jack built.'];
        }

        let string = 'This is ' + this.recVerse(verseNumber);
        return string.split('\n');
    }

    static verses(startVerse, endVerse) {
        let versesList = [];

        for (let i = startVerse; i <= endVerse; i++) {
            versesList.push(...this.verse(i));
            if (i < endVerse) {
                versesList.push('');
            }
        }

        return versesList;
    }

    static recVerse(number) {
        if(number === 2){
            return this.items[number].name 
                + '\n' + this.items[number].action;
        }

        return this.items[number].name 
            + '\nthat ' + this.items[number].action 
            + ' ' + this.recVerse(number - 1);
    }
}
